#1222 Standardising and sharing lockscreen code, with added debug logging for non-IOS

drmason789 opened the pull request:
As a first phase for #1222 I have moved the lockscreen code that is repeated in many classes, into a small set of shared methods.

- Code to detect and set up the Peloton workaround is moved to the bluetoothdevice constructor
- Code to update to and from the lockscreen is moved to methods in bluetoothdevice class.
- Code to implement the Peloton workaround is accessed via an abstract method in bluetoothdevice with implemetnations in the next level of subclasses.
- The various "firstInit", "firstVirtualThing", "firstStateChanged" etc that are used to flag if the virtual device setup has been done has been moved to class bluetoothdevice, so it's consistent across all devices and also available for some shared lockscreen code.
- The new methods for lockscreen code are now called in non-IOS builds but they do nothing. In a second phase of #1222, these methods will call methods on an object that would be replaceable with one that can, in non-IOS builds, record if the expected lockscreen functions have been called, in a unit testing scenario.

drmason789 commented:
@cagnulein This isn't ready for merging...I'm going to have a few questions about this.

drmason789 commented on src/bluetoothdevice.h:
```diff
@@ -417,9 +452,25 @@ class bluetoothdevice : public QObject {
 
+  private:
+
+    bool virtualDeviceSetupDone = false;
```
This replaces the "firstInit", "firstVirtualX", "firstStateChanged" flags used to determine (in most cases) if the virtual device setup has been done before.

cagnulein commented:
thanks @drmason789 ! let me know when it should be ready for merge!

drmason789 commented on src/bowflext216treadmill.cpp:
```diff
@@ -92,14 +92,14 @@ void bowflext216treadmill::update() {
         QSettings settings;
         // ******************************************* virtual treadmill init *************************************
-        if (!firstInit && !virtualTreadMill) {
+        if (!this->isVirtualDeviceSetUp() && !virtualTreadMill) {
```
No lockscreen/Peloton workaround check.

drmason789 replied:
No lockscreen updates at all in this class.

drmason789 commented on src/rower.cpp:
```diff
@@ -3,7 +3,24 @@
+
+            qDebug() << "ios_peloton_workaround activated!";
+            this->lockScreen = new lockscreen();
+            this->lockScreen->virtualbike_ios();
```
It's a rower, but uses the virtual bike lockscreen setup. Copied from one of the rower classes.

drmason789 commented on src/rower.cpp:
```diff
@@ -39,6 +56,25 @@ void rower::powerSensor(uint16_t power) { m_watt.setValue(power, false); }
+    bool ios_peloton_workaround = settings.value(QZSettings::ios_peloton_workaround, QZSettings::default_ios_peloton_workaround).toBool();
+    bool virtual_device_rower = settings.value(QZSettings::virtual_device_rower, QZSettings::default_virtual_device_rower).toBool();
+    if (ios_peloton_workaround && cadence && !virtual_device_rower) {
+        this->lockScreen->virtualbike_setCadence(currentCrankRevolutions(), lastCrankEventTime());
```
bike methods for a rower.

drmason789 commented on src/soleelliptical.cpp:
```diff
@@ -144,7 +144,7 @@ void soleelliptical::update() {
 
         // ******************************************* virtual treadmill init *************************************
-        if (!firstVirtual && searchStopped && !virtualTreadmill && !virtualBike) {
+        if (!this->isVirtualDeviceSetUp() && searchStopped && !virtualTreadmill && !virtualBike) {
```
No lockscreen/Peloton workaround check.

drmason789 commented on src/soleelliptical.cpp:
```diff
@@ -163,7 +163,7 @@ void soleelliptical::update() {
                             &soleelliptical::changeInclinationRequested);
                 }
-                firstVirtual = 1;
+                this->setVirtualDeviceSetUp();
```
To line 169, I think.

drmason789 commented on src/spirittreadmill.cpp:
```diff
@@ -443,7 +443,7 @@ void spirittreadmill::stateChanged(QLowEnergyService::ServiceState state) {
 
         // ******************************************* virtual treadmill init *************************************
-        if (!firstVirtualTreadmill && !virtualTreadMill && !virtualBike) {
+        if (!this->isVirtualDeviceSetUp() && !virtualTreadMill && !virtualBike) {
```
No lockscreen/Peloton workaround check.

drmason789 commented on src/sportsplusbike.cpp:
```diff
@@ -357,7 +357,7 @@ void sportsplusbike::stateChanged(QLowEnergyService::ServiceState state) {
 
         // ******************************************* virtual bike init *************************************
-        if (!firstVirtualBike && !virtualBike) {
+        if (!this->isVirtualDeviceSetUp() && !virtualBike) {
```
No lockscreen/Peloton workaround check.

drmason789 commented on src/sportstechbike.cpp:
```diff
@@ -275,7 +275,7 @@ void sportstechbike::stateChanged(QLowEnergyService::ServiceState state) {
 
         // ******************************************* virtual bike init *************************************
-        if (!firstVirtualBike && !virtualBike) {
+        if (!this->isVirtualDeviceSetUp() && !virtualBike) {
```
No lockscreen/Peloton workaround check.

drmason789 commented on src/strydrunpowersensor.cpp:
```diff
@@ -272,32 +271,13 @@ void strydrunpowersensor::characteristicChanged(const QLowEnergyCharacteristic &
     }
 
+
     if (!noVirtualDevice) {
```
I'm not yet worked out why this condition is in this class, but not most others.

drmason789 commented on src/technogymmyruntreadmillrfcomm.cpp:
```diff
@@ -103,6 +103,9 @@ void technogymmyruntreadmillrfcomm::update() {
+
+        // TODO: determine why no checking if this has been done before
+
         if (!virtualTreadMill) {
```
@cagnulein Why no check if the virtual device setup or "first state changed" here?

cagnulein replied:
there is only one use that is using this, so it's a working in progress module

drmason789 commented on src/technogymmyruntreadmillrfcomm.cpp:
```diff
@@ -103,6 +103,9 @@ void technogymmyruntreadmillrfcomm::update() {
 
```
No lockscreen functionality in this class.

drmason789 commented on src/toorxtreadmill.cpp:
```diff
@@ -79,6 +79,9 @@ void toorxtreadmill::update() {
+
+        // TODO: determine why no checking if this has been done before
+
         if (!virtualTreadMill) {
```
No check of if virtual device setup has been done before
No locksreen/Peloton workaround check.
No lockscreen functionality in this class.

cagnulein replied:
yes this module it's not used (at least I guess)

drmason789 commented on src/treadmill.cpp:
```diff
@@ -164,3 +164,38 @@ void treadmill::groundContactSensor(double groundContact) { GroundContactMS.setV
+    bool ios_peloton_workaround =
+            settings.value(QZSettings::ios_peloton_workaround, QZSettings::default_ios_peloton_workaround).toBool();
+    if (ios_peloton_workaround && cadence) {
+        this->lockScreen->virtualbike_setCadence(currentCrankRevolutions(), lastCrankEventTime());
```
bike functions for a treadmill

drmason789 reviewed (commented):
Overall,
- I believe we need to fix the location of the setVirtualDeviceSetUp() in many cases because if the virtual device is disabled, the setup will keep repeating until it isn't.
- The lockscreen functionality that is now shared is specifying lockscreen functions for  bikes for non-bikes, etc. Could we confirm these are correct and justify unexpected mismatches?
- The lockscreen update code is missing from some classes
- The Peloton workaround is missing from some classes.

drmason789 commented:
@cagnulein Ready for your review.

drmason789 commented on src/bluetoothdevice.cpp:
```diff
@@ -278,11 +286,40 @@ QStringList bluetoothdevice::metrics() {
+    lockscreen h;
+    long appleWatchHeartRate = h.heartRate();
+    this->Heart = appleWatchHeartRate;
+    qDebug() << "Current Heart from Apple Watch: " << QString::number(appleWatchHeartRate);
```
Changed from using the subclass' debug signal because that signal is not defined at this level.

drmason789 commented on src/bluetoothdevice.cpp:
```diff
@@ -278,11 +286,40 @@ QStringList bluetoothdevice::metrics() {
+    h.setKcal(KCal.value());
+    h.setDistance(Distance.value());
+    this->Heart = appleWatchHeartRate!=0 ? appleWatchHeartRate:defaultHeartRate;
+    qDebug() << "Current Heart from Apple Watch: " << QString::number(appleWatchHeartRate);
```
Changed from using the subclass' debug signal because that signal is not defined at this level.

cagnulein commented on src/activiotreadmill.cpp:
```diff
@@ -183,7 +181,7 @@ void activiotreadmill::update() {
                             &activiotreadmill::changeInclinationRequested);
                 }
-                firstInit = 1;
+                this->setVirtualDeviceSetUp();
```
yes right

cagnulein commented on src/virtualrower.cpp:
```diff
@@ -10,6 +10,26 @@
+
+        qDebug() << "ios_zwift_workaround activated!";
+        this->lockScreen = new lockscreen();
+        this->lockScreen->virtualrower_ios();
```
that's because virtualrower has one only interface, so this is the right one for zwift and everything

cagnulein commented on src/elliptical.cpp:
```diff
@@ -2,7 +2,19 @@
+        if (ios_peloton_workaround && cadence) {
+            qDebug() << "ios_peloton_workaround activated!";
+            h = new lockscreen();
+            h->virtualbike_ios();
```
yes it-s a special case. take the domyos elliptical as a reference

drmason789 replied:
Domyos Elliptical doesn't have the Peloton workaround - I've added it in this PR and it's inheriting the same code.

https://github.com/cagnulein/qdomyos-zwift/blob/5d1b399fa6de4a146a7c0f721721ce8583c40f4d/src/domyoselliptical.cpp#L171-L193

cagnulein commented on src/eslinkertreadmill.cpp:
```diff
@@ -110,13 +110,13 @@ void eslinkertreadmill::update() {
                 virtualTreadMill = new virtualtreadmill(this, noHeartService);
                 connect(virtualTreadMill, &virtualtreadmill::debug, this, &eslinkertreadmill::debug);
-                firstInit = 1;
+                this->setVirtualDeviceSetUp();
```
yes indeed

cagnulein commented on src/eslinkertreadmill.cpp:
```diff
@@ -110,13 +110,13 @@ void eslinkertreadmill::update() {
         QSettings settings;
         // ******************************************* virtual treadmill init *************************************
-        if (!firstInit && !virtualTreadMill) {
+        if (!this->isVirtualDeviceSetUp() && !virtualTreadMill) {
```
this treadmill is not supported/used, but yes we can match with the other implementation

cagnulein commented on src/fakeelliptical.cpp:
```diff
@@ -45,37 +44,21 @@ void fakeelliptical::update() {
+
+        if (virtual_device_enabled) {
             emit debug(QStringLiteral("creating virtual bike interface..."));
             virtualBike = new virtualbike(this, noWriteResistance, noHeartService);
```
Yeah as I said elliptical can be both treadmill and bike like the domyos elliptical

cagnulein commented on src/fakeelliptical.cpp:
```diff
@@ -45,37 +44,21 @@ void fakeelliptical::update() {
     }
-    if (!firstStateChanged)
+    if (!this->isVirtualDeviceSetUp())
         emit connectedAndDiscovered();
```
hah right, it's because fake devices aren't about a bluetooth device so that signal is required to submit to the main bluetooth class their presence

drmason789 replied:
I added some explanation in comments.

cagnulein commented on src/faketreadmill.cpp:
```diff
@@ -71,9 +70,9 @@ void faketreadmill::update() {
         }
-        if (!firstStateChanged)
+        if (!this->isVirtualDeviceSetUp())
             emit connectedAndDiscovered();
```
yep same as above

cagnulein commented on src/faketreadmill.cpp:
```diff
@@ -52,7 +51,7 @@ void faketreadmill::update() {
 
     // ******************************************* virtual treadmill init *************************************
-    if (!firstStateChanged && !virtualTreadmill && !virtualBike) {
+    if (!this->isVirtualDeviceSetUp() && !virtualTreadmill && !virtualBike) {
```
yes we can align to the other ones

cagnulein commented on src/faketreadmill.cpp:
```diff
@@ -71,9 +70,9 @@ void faketreadmill::update() {
+        if (!this->isVirtualDeviceSetUp())
             emit connectedAndDiscovered();
-        firstStateChanged = 1;
+        this->setVirtualDeviceSetUp();
```
yep

cagnulein commented on src/fitshowtreadmill.cpp:
```diff
@@ -47,10 +44,6 @@ fitshowtreadmill::~fitshowtreadmill() {
     }
-#if defined(Q_OS_IOS) && !defined(IO_UNDER_QT)
-    if (h)
-        delete h;
```
ok

cagnulein commented on src/fitshowtreadmill.cpp:
```diff
@@ -171,7 +164,7 @@ void fitshowtreadmill::update() {
                         &fitshowtreadmill::changeInclinationRequested);
 
-                firstInit = 1;
+                this->setVirtualDeviceSetUp();
```
yep

cagnulein commented on src/m3ibike.cpp:
```diff
@@ -759,11 +726,11 @@ void m3ibike::processAdvertising(const QByteArray &data) {
 }
 
-bool m3ibike::connected() { return initDone; }
+bool m3ibike::connected() { return this->isVirtualDeviceSetUp(); }
```
yes it's better because this is a special device. it's better to keep these separated

cagnulein commented on src/nautilusbike.cpp:
```diff
@@ -91,13 +91,13 @@ void nautilusbike::update() {
         QSettings settings;
         // ******************************************* virtual treadmill init *************************************
-        if (!firstVirtual && !virtualBike) {
+        if (!this->isVirtualDeviceSetUp() && !virtualBike) {
```
yeah need to be aligned

cagnulein commented on src/nautilusbike.cpp:
```diff
@@ -91,13 +91,13 @@ void nautilusbike::update() {
                 virtualBike = new virtualbike(this);
                 connect(virtualBike, &virtualbike::changeInclination, this, &nautilusbike::changeInclinationRequested);
-                firstVirtual = 1;
+                this->setVirtualDeviceSetUp();
```
yes right

cagnulein commented on src/nautiluselliptical.cpp:
```diff
@@ -138,7 +138,7 @@ void nautiluselliptical::update() {
                             &nautiluselliptical::changeInclinationRequested);
                 }
-                firstVirtual = 1;
+                this->setVirtualDeviceSetUp();
```
right

cagnulein commented on src/nordictrackelliptical.cpp:
```diff
@@ -827,7 +817,7 @@ void nordictrackelliptical::stateChanged(QLowEnergyService::ServiceState state)
                             &nordictrackelliptical::changeInclinationRequested);
                 }
-                firstStateChanged = 1;
+                this->setVirtualDeviceSetUp();
```
right

cagnulein commented on src/nordictrackifitadbbike.cpp:
```diff
@@ -33,15 +32,15 @@ nordictrackifitadbbike::nordictrackifitadbbike(bool noWriteResistance, bool noHe
             connect(virtualBike, &virtualbike::changeInclination, this,
                     &nordictrackifitadbbike::changeInclinationRequested);
-            firstStateChanged = 1;
+            this->setVirtualDeviceSetUp();
```
yep

cagnulein commented on src/nordictrackifitadbtreadmill.cpp:
```diff
@@ -54,7 +53,7 @@ nordictrackifitadbtreadmill::nordictrackifitadbtreadmill(bool noWriteResistance,
                         &nordictrackifitadbtreadmill::changeInclinationRequested);
             }
-            firstStateChanged = 1;
+            this->setVirtualDeviceSetUp();
```
yep

cagnulein commented on src/octaneelliptical.cpp:
```diff
@@ -260,7 +260,7 @@ void octaneelliptical::update() {
         QSettings settings;
         // ******************************************* virtual treadmill init *************************************
-        if (!firstInit && !virtualTreadMill) {
+        if (!this->isVirtualDeviceSetUp() && !virtualTreadMill) {
```
need to be aligned

cagnulein commented on src/octaneelliptical.cpp:
```diff
@@ -269,7 +269,7 @@ void octaneelliptical::update() {
                 connect(virtualTreadMill, &virtualtreadmill::changeInclination, this,
                         &octaneelliptical::changeInclinationRequested);
-                firstInit = 1;
+                this->setVirtualDeviceSetUp();
```
yes

cagnulein commented on src/octanetreadmill.cpp:
```diff
@@ -266,15 +266,15 @@ void octanetreadmill::update() {
         QSettings settings;
         // ******************************************* virtual treadmill init *************************************
-        if (!firstInit && !virtualTreadMill) {
+        if (!this->isVirtualDeviceSetUp() && !virtualTreadMill) {
```
needs to be aligned

cagnulein commented on src/paferstreadmill.cpp:
```diff
@@ -112,7 +112,7 @@ void paferstreadmill::update() {
         QSettings settings;
         // ******************************************* virtual treadmill init *************************************
-        if (!firstInit && !virtualTreadMill) {
+        if (!this->isVirtualDeviceSetUp() && !virtualTreadMill) {
```
needs to be aligned yes

cagnulein commented on src/paferstreadmill.cpp:
```diff
@@ -121,7 +121,7 @@ void paferstreadmill::update() {
                 connect(virtualTreadMill, &virtualtreadmill::changeInclination, this,
                         &paferstreadmill::changeInclinationRequested);
-                firstInit = 1;
+                this->setVirtualDeviceSetUp();
```
yes

cagnulein commented on src/pelotonbike.cpp:
```diff
@@ -25,14 +24,14 @@ pelotonbike::pelotonbike(bool noWriteResistance, bool noHeartService) {
             virtualBike = new virtualbike(this);
             connect(virtualBike, &virtualbike::changeInclination, this, &pelotonbike::changeInclinationRequested);
-            firstStateChanged = 1;
+            this->setVirtualDeviceSetUp();
```
yes

cagnulein commented on src/proformelliptical.cpp:
```diff
@@ -394,7 +384,7 @@ void proformelliptical::stateChanged(QLowEnergyService::ServiceState state) {
         // ******************************************* virtual treadmill init *************************************
         QSettings settings;
-        if (!firstStateChanged && !virtualTreadmill && !virtualBike) {
+        if (!this->isVirtualDeviceSetUp() && !virtualTreadmill && !virtualBike) {
```
need to be aligned

cagnulein commented on src/proformelliptical.cpp:
```diff
@@ -410,7 +400,7 @@ void proformelliptical::stateChanged(QLowEnergyService::ServiceState state) {
                             &proformelliptical::changeInclinationRequested);
                 }
-                firstStateChanged = 1;
+                this->setVirtualDeviceSetUp();
```
yes

cagnulein commented on src/proformellipticaltrainer.cpp:
```diff
@@ -567,7 +557,7 @@ void proformellipticaltrainer::stateChanged(QLowEnergyService::ServiceState stat
         // ******************************************* virtual treadmill init *************************************
         QSettings settings;
-        if (!firstStateChanged && !virtualTreadmill && !virtualBike) {
+        if (!this->isVirtualDeviceSetUp() && !virtualTreadmill && !virtualBike) {
```
need to be aligned

cagnulein commented:
@drmason789 there is an important thing about virtual devices creation: the placement inside the code matters. I have some cases on android where the creation of the virtual devices it's really important (for example creating before receiving the description written characteristic could be an issue). So the costructor of the virtualbike/treadmil.. must remain in the same linecode for each module, of course you can create a single function for each ones as you're already doing. I don't know if it's clear

cagnulein commented:
i guess I commented all of them. In general you were right about the fact that sometimes the `setVirtualDeviceSetUp` needs to be moved out of the if() thing. and also all the classes must have the `lockscreen ` implementation.

drmason789 commented:
Plan for adding Peloton workaround and other lockscreen code:
- Add the !lockscreen condition to the if statement that checks for virtual device setup
- at the end of the characteristicChanged method, call updateLockscreenEnergyDistanceHeartRate() if the heart rate belt is disabled in the settings
- at the end of characteristicChanged, call doPelotonWorkaround() 

Classes where it's missing:
bowflext216treadmill
bowflextreadmill
domyosrower
eliterizer
elitesterzosmart
eslinkertreadmill
fitmetria_fanfit
heartratebelt
iconceptbike
nautilusbike
nautiluselliptical
nautilustreadmill
octaneelliptical
octanetreadmill
paferstreadmill
smartspin2k
spirittreadmill
sportsplusbike
sportstechbike
technogymmyruntreadmillrfcomm
toorxtreadmill

drmason789 commented:
@cagnulein For the update of the energy, distance, and heart rate to and from the lockscreen, should the Android keepawake alternative be in every bluetoothdevice class too?

i.e. this bit:
```
#ifdef Q_OS_ANDROID
    if (settings.value(QZSettings::ant_heart, QZSettings::default_ant_heart).toBool())
        Heart = (uint8_t)KeepAwakeHelper::heart();
    else
#endif
    {
        if (heartRateBeltName.startsWith(QStringLiteral("Disabled"))) {
            this->updateLockscreenEnergyDistanceHeartRate();
        }
    }
```

drmason789 commented:
Additionally, for the likes of:

```
/*if(heartRateBeltName.startsWith("Disabled"))
        Heart = value.at(18);*/
```

I will replace the code, but leave it commented out:

```
        /*
        if(heartRateBeltName.startsWith("Disabled"))
            if(!this->updateLockscreenEnergyDistanceHeartRate())
                this->Heart = value.at(18);
        */
```

I.e. if this was uncommented, if the HRB is disabled, try to do the lockscreen update (which may in the future call a test stub on a non IOS build) and if that reports it hasn't done the change, get the heart rate from the data packet.

cagnulein commented:
> I'm not sure if lockscreen / Peloton workaround code belongs in these:
> eliterizer
> elitesterzosmart
> fitmetria_fanfit
> heartratebelt
> smartspin2k

confirmed, no need to add lockscreen/peloton here

I will wait this PR before doing #1242 @rickyny

drmason789 commented:
> @drmason789 there is an important thing about virtual devices creation: the placement inside the code matters. I have some cases on android where the creation of the virtual devices it's really important (for example creating before receiving the description written characteristic could be an issue). So the costructor of the virtualbike/treadmil.. must remain in the same linecode for each module, of course you can create a single function for each ones as you're already doing. I don't know if it's clear

When you say "virtual devices" do you specifically mean the creation of the virtualbike, virtualtreadmill, virtualrower classes (which I haven't moved) or something that's created in the IOS code from calling  lockscreen::virtual* functions , which I have moved to the bike, elliptical, treadmill and rower constructors ?

drmason789 commented on src/kingsmithr2treadmill.cpp:
```diff
@@ -309,17 +308,7 @@ void kingsmithr2treadmill::characteristicChanged(const QLowEnergyCharacteristic
 
             uint8_t heart = 0;
             if (heart == 0) {
-
```
No heart rate except via IOS lockscreen. I suspect this is a bug, but unless you suggest alternative behaviour I will just remap to the new function.

cagnulein replied:
yeah i did that on purpose only to remember me, that if, in the future, i have to add the HR from the machinery, that will be the point to add. So it's ok that it's always going into the iOS code

drmason789 commented on src/kingsmithr1protreadmill.cpp:
```diff
@@ -303,17 +302,7 @@ void kingsmithr1protreadmill::characteristicChanged(const QLowEnergyCharacterist
 
             uint8_t heart = 0;
             if (heart == 0) {
-
```
No heart rate except via IOS lockscreen. I suspect this is a bug, but unless you suggest alternative behaviour I will just remap to the new function.

cagnulein replied:
yeah i did that on purpose only to remember me, that if, in the future, i have to add the HR from the machinery, that will be the point to add. So it's ok that it's always going into the iOS code

drmason789 commented:
@cagnulein This may be mergeable currently. However, I intend to do another pass of looking at each bluetoothdevice subclass, looking at what it did before for Peloton workaround and energy/distance/heart update and step cadence update, and if this is consistent with what it now inherits from its superclass constructor and doPelotonWorkaround override.

cagnulein commented:
ok let's to merge this now (i have time tomorrow i guess). And then we can go on, is it ok for you?

drmason789 commented:
@cagnulein I believe it is easier for you to look at the code that is now inherited and decide if that is correct for the classes that inherit it. I need to look at the individual classes, what they did before, compare with what they now inherit, and if there is a mismatch (eg rower using bike functionality) check with you. So if you are happy to test and merge this and move on, so am I.

cagnulein commented:
@drmason789 totally right and thanks for all you're doing, you did an amazing work here!

drmason789 commented:
@cagnulein Thanks. I think it's a great thing you're doing for the fitness community.

drmason789 commented:
@cagnulein I've done the second phase of #1222 in PR#1262. I would prefer to simply merge that into this PR or abandon this one and continue with that.

At this point, it's not set up to run a bluetoothdevice object and test that it's calling the lockscreen functions on non IOS platforms using a test stub implementation, but I wondered if you think a fake lockscreen (debuglockscreen?) alternative that simply logs calls and state via qdebug would be helpful. This could be enabled in the settings, or simply assumed when debug log is enabled, then you would see in the log if the lockscreen functions, including the IOS peloton workaround, if enabled, are being called, even in non IOS.

cagnulein commented:
> I've done the second phase of #1222 in PR#1262. I would prefer to simply merge that into this PR or abandon this one and continue with that.

yes at this point it makes sense to do a merge in one step, I agree.

> At this point, it's not set up to run a bluetoothdevice object and test that it's calling the lockscreen functions on non IOS platforms using a test stub implementation, but I wondered if you think a fake lockscreen (debuglockscreen?) alternative that simply logs calls and state via qdebug would be helpful. This could be enabled in the settings, or simply assumed when debug log is enabled, then you would see in the log if the lockscreen functions, including the IOS peloton workaround, if enabled, are being called, even in non IOS.

yes it makes sense, we can enable it when the debug log is enabled, right, so you can directly check yourself is the code is doing the right thing!

drmason789 commented:
I see in #995 :

>>@cagnulein Throughout the device code, where the virtual devices are created:
>>
>>The iOS homescreen code for the iOS Peloton workaround is in some devices, but not others. Should it be in all devices?
>>
>only on the bikes

I perceive there to be a distinction between lockscreen updates and the Peloton workaround. These can be separately active.

In this PR (done much later) I've put the Peloton workaround on all devices except for a few that are really peripherals.

When I ask "should it be on all devices" and you say "only on the bikes" I would interpret this to mean "the Peloton workaround should only be implemented on the bikes, not the others".

Could you confirm this, in which case I will remove it from the non-bikes.

As it is implemented in this PR, we could leave the calls in the other classes, but have them do nothing in the non-bikes. Then in the future, if you want a Peloton workaround for the other devices, it could potentially be inserted via a method override in the immediate subclasses of bluetoothdevice.

drmason789 commented:
@cagnulein I notice that amongst the treadmills, only the truetreadmill does the step cadence lockscreen update. Should all the treadmills do this?

drmason789 commented:
I've done a review of the bluetoothdevice subclasses, paying attention to what they did before wrt Peloton workaround and the other lockscreen update. I've identified some changes in behaviour.

Some of these changes in behaviour include the addition of the Peloton workaround because while the Peloton workaround update code is present, the lockscreen object isn't constructed.

**Legend**
PT = added Peloton Workaround for Treadmills, defined in treadmill::doPelotonWorkaround()
PR = added Peloton Workaround for Rowers, defined in rower::doPelotonWorkaround()
PE = added Peloton Workaround for Elliptical Trainers, defined in elliptical::doPelotonWorkaround()
PB = added Peloton Workaround for Bikes, defined in bike::doPelotonWorkaround()
EDH = added Energy, Distance and Heartrate update to/from lockscreen

**Changes**

```
activiotreadmill:	PT
bhfitnesselliptical:	PE
bowflext216treadmill:	PT
bowflextreadmill:	PT
concept2skierg:	PR from PB
domyoselliptical:	PE
domyosrower:	PR, EDH
domyostreadmill:	PT
echelonrower:	PR from PB
echelonstride:	PT
eslinkertreadmill:	PT, EDH
fakeelliptical:	PE from PB
faketreadmill:	PT
fitshowtreadmill:	PT
ftmsrower:	PR from PB
horizontreadmill:	PT
iconceptbike:	PB
kingsmithr1protreadmill:	PT
kingsmithr2treadmill:	PT
lifefitnesstreadmill:	PT
m3ibike:	PB
nautilusbike:	PB,EDH
nautiluselliptical:	PE,EDH
nautilustreadmill:	PT
nordictrackelliptical:	PE
nordictrackifitadbbike:	PB
nordictrackifitadbtreadmill:	PT
octaneelliptical:	PE
octanetreadmill:	PT
paferstreadmill:	PT
pelotonbike:	PB
proformelliptical:	PE
proformellipticaltrainer:	PE
proformrower:	PR
proformtreadmill:	PT
proformwifitreadmill:	PT
shuaa5treadmill:	PT
skandikawiribike:	PB,EDH
smartrowrower:	PR from PB
soleelliptical:	PE
solef80treadmill:	PT
spirittreadmill:	PT
sportsplusbike:	PB, EDH
sportstechbike:	PB, EDH
strydrunpowersensor:	PT
technogymmyruntreadmill:	PT
technogymmyruntreadmillrfcomm:	PT,EDH
toorxtreadmill:	PT,EDH
truetreadmill:	PT
trxappgateusbbike:	PB
trxappgateusbtreadmill:	PT
```

drmason789 commented:
The remaining questions I have:
- Should the non-bikes have a Peloton workaround? Some of them do already, such as Echelon rower, Smartrow rower, FTMS rower
- Should all the treadmills have the step cadence update?
- If all the devices should have the Peloton workaround, should the functions called on the lockscreen object be virtualbike for bike, virtualtreadmill for treadmills (and maybe ellipticals), and virtualrower for rowers? Currently a few rowers are using the virtualbike lockscreen functions.

drmason789 commented:
Additionally, had #995 been merged before this, I would have shared some of the virtual(bike|treadmill|rower) functionality in the virtualdevice class the PR introduces. Would you object to me merging that PR into this one prior to testing and making this adjustment?

cagnulein commented:
> I see in #995 :
> 
> > > @cagnulein Throughout the device code, where the virtual devices are created:
> > > The iOS homescreen code for the iOS Peloton workaround is in some devices, but not others. Should it be in all devices?
> > 
> > 
> > only on the bikes
> 
> I perceive there to be a distinction between lockscreen updates and the Peloton workaround. These can be separately active.
> 
> In this PR (done much later) I've put the Peloton workaround on all devices except for a few that are really peripherals.
> 
> When I ask "should it be on all devices" and you say "only on the bikes" I would interpret this to mean "the Peloton workaround should only be implemented on the bikes, not the others".
> 
> Could you confirm this, in which case I will remove it from the non-bikes. But some rowers had it before this PR.
> 
> As it is implemented in this PR, we could leave the calls in the other classes, but have them do nothing in the non-bikes. Then in the future, if you want a Peloton workaround for the other devices, it could potentially be inserted via a method override in the immediate subclasses of bluetoothdevice.

I guess I said it wrong.

the lockscreen thing was in order to update the apple watch thing. The ios peloton workaround is to fix a QT issue on iOS that leads to a crash.

So both of them are need for all the kind of the devices actually.

cagnulein commented:
ok i checked the code and now i understood because you can't find the ios peloton workaround in the treadmill. because it's in the virtual treadmill class! maybe the right solution would be put the ios peloton workaround in the virtualbike as well?

cagnulein commented:
hah I got now because only the bikes has the ios peloton workaround in the device classes: because for peloton i have to send cadence and hr to swift interface. so it makes sense that only on bikes you have this. So final answer: the ios peloton workaround for cadence and hr is ok to have only on bike.

cagnulein commented:
> * Should all the treadmills have the step cadence update? Currently only the truetreadmill has it.

also the domyostreadmill has it, but yes all the treadmill should have this! thanks!

cagnulein commented:
> * If all the devices should have the Peloton workaround, should the functions called on the lockscreen object be virtualbike for bike, virtualtreadmill for treadmills (and maybe ellipticals), and virtualrower for rowers? Currently a few rowers are using the virtualbike lockscreen functions.

I guess I already answered to this, some comment before, let me know if it was clear or not

drmason789 commented:
With my last commit, I have changed the call to doPelotonWorkaround() all the classes now do, to a more general doLockscreenUpdate(). This is overridden in bike to call doPelotonWorkaround(). In treadmill, it is overridden to update the step cadence. So that covers "only do the Peloton workaround in the device class, for the bikes" and "all treadmills should have the step cadence update".

drmason789 commented:
Then there are these ones...non-bikes that had a bike Peloton update, which I had changed to the Peloton update for the specific device, which is now only done for the bikes. So currently these devices are all Px to none.

```
concept2skierg:	PR from PB
echelonrower:	PR from PB
fakeelliptical:	PE from PB
ftmsrower:	PR from PB
smartrowrower:	PR from PB
```

However, they have some virtual device mismatches:

```
concept2skierg:	rower using virtualtreadmill
echelonrower:	rower using virtualbike
fakeelliptical:	elliptical using virtualbike
ftmsrower:	rower using virtualbike
smartrowrower:	rower using virtualbike
```

drmason789 commented:
When the Peloton workaround is actually being used in IOS, the virtualbike would not be created. Should the rule perhaps be that the Peloton workaround should happen for bluetoothdevice subclasses that normally use a virtualbike?

There are also those cases where a different virtual device is used depending on settings.

drmason789 commented:
> > * If all the devices should have the Peloton workaround, should the functions called on the lockscreen object be virtualbike for bike, virtualtreadmill for treadmills (and maybe ellipticals), and virtualrower for rowers? Currently a few rowers are using the virtualbike lockscreen functions.
> 
> I guess I already answered to this, some comment before, let me know if it was clear or not

Agreed, you already answered this.

cagnulein commented:
sorry @drmason789 if I didn't merge this yet, i can't keep the pace of all the ticket and mail that I'm receiving :(

cagnulein commented:
sorry which unanswered one? Probably i lost them. I thought i answered to everyone 2 days ago

drmason789 commented:
I used your answers to get to commit 91B3446, then I asked more refined questions.

cagnulein commented:
> When the Peloton workaround is actually being used in IOS, the virtualbike would not be created. Should the rule perhaps be that the Peloton workaround should happen for bluetoothdevice subclasses that normally use a virtualbike?

nope. Virtual bike is not created at all ONLY if `ios peloton workaround` is enabled AND `bike_cadence_sensor` is enabled as well.

In the other cases, virtualbike is created, you can check this as reference https://github.com/cagnulein/qdomyos-zwift/blob/f81f5354f665f247c2223c93597826dd685f02e0/src/echelonconnectsport.cpp#L380

Let me know if you have any other questions

drmason789 commented:
> > When the Peloton workaround is actually being used in IOS, the virtualbike would not be created. Should the rule perhaps be that the Peloton workaround should happen for bluetoothdevice subclasses that normally use a virtualbike?
> 
> nope. Virtual bike is not created at all ONLY if `ios peloton workaround` is enabled AND `bike_cadence_sensor` is enabled as well.
> 
> In the other cases, virtualbike is created, you can check this as reference
> 
> https://github.com/cagnulein/qdomyos-zwift/blob/f81f5354f665f247c2223c93597826dd685f02e0/src/echelonconnectsport.cpp#L380
> 
> Let me know if you have any other questions

I meant the rule to be  "When the Peloton workaround is actually being used in IOS", i.e. `ios peloton workaround` is enabled AND `bike_cadence_sensor` is enabled as well.

Also, I'm asking about the non-bikes that were doing the Peloton update for bikes, unlike the echeloconectsport which is a bike. I think I've found my answer amongst the conditions for creating the lockscreen object or the virtualbike.

I am still not ready for this to be merged, until I've confirmed I've matched behaviour with the previous implementation for the non-bikes.

drmason789 commented on src/echelonrower.cpp:
```diff
@@ -256,30 +255,12 @@ void echelonrower::characteristicChanged(const QLowEnergyCharacteristic &charact
-    bool ios_peloton_workaround = settings.value(QZSettings::ios_peloton_workaround, QZSettings::default_ios_peloton_workaround).toBool();
-    bool virtual_device_rower = settings.value(QZSettings::virtual_device_rower, QZSettings::default_virtual_device_rower).toBool();
-    if (ios_peloton_workaround && cadence && !virtual_device_rower && h && firstStateChanged) {
-        h->virtualbike_setCadence(currentCrankRevolutions(), lastCrankEventTime());
```
Based on later answers, this is because when !virtual)device_rower (in addition to the usual conditions) this device emulates a bike rather than a rower, so needs this update.

drmason789 commented:
If I define the "standard conditions" for using the Peloton workaround update with a bike as: 
    ios_peloton_workaround and bike_cadence_sensor
and "extra conditions" as additional conditions to these for using the Peloton workaround update:

**concept2skierg:	rower using virtualtreadmill**
This one, outside of IOS, uses a virtualtreadmill, so no Peloton workaround.
But in IOS, uses lockscreen as a bike with no extra conditions.
* this seems like an inconsistency, it's only being a bike for the Peloton workaround under IOS.

**echelonrower:	rower using virtualbike**
Outside of IOS, if virtual_device_rower this uses a virtualrower, otherwise uses a virtualbike.
In IOS, if !virtual_device_rower in addition to the standard conditions, uses the workaround.
* OK, it's emulating a bike, so use the Peloton workaround for bike

**fakeelliptical:	elliptical using virtualbike
ftmsrower:	rower using virtualbike
smartrowrower:	rower using virtualbike**
Outside of IOS, these rowers always use a virtual bike. 
In IOS, they do the bike Peloton workaround, with no extra conditions.

**Others, e.g. proformrower**
Outside of IOS, if virtual_device_enabled && !virtual_device_rower && virtual_device_force_bike it uses a virtualbike
Bit in IOS, there's no Peloton workaround update even when the standard conditions are met and it would be emulating a bike outside of IOS.

cagnulein commented:
> **concept2skierg: rower using virtualtreadmill** This one, outside of IOS, uses a virtualtreadmill, so no Peloton workaround. But in IOS, uses lockscreen as a bike with no extra conditions.
> 
> * this seems like an inconsistency, it's only being a bike for the Peloton workaround under IOS.

yes true, I guess no one uses that module and we should ask to the OP of that trainer how he uses it. (if he still uses it )



> **echelonrower: rower using virtualbike** Outside of IOS, if virtual_device_rower this uses a virtualrower, otherwise uses a virtualbike. In IOS, if !virtual_device_rower in addition to the standard conditions, uses the workaround.
> 
> * OK, it's emulating a bike, so use the Peloton workaround for bike

these strange cases derived from the user requests. Sometime a users asked me to add a virtualbike capabilities to a rower because he likes to play it on zwift like a bike.

As a general behaviour, we can think this, in this way: all the kind of trainer could create a virtual rower, treadmiill, or bike (only one of them at a time) selected by the virtual_device_rower settig, virtual_treadmill and so on). What do you think? Does this simplify the logic?

cagnulein commented:
hah of course if a device it's a treadmill default virtual device will be a treadmill, if it's a bike, virtual bike is the default and so on. for elliptical virtual treadmill is the default

drmason789 commented:
> As a general behaviour, we can think this, in this way: all the kind of trainer could create a virtual rower, treadmiill, or bike (only one of them at a time) selected by the virtual_device_rower settig, virtual_treadmill and so on). What do you think? Does this simplify the logic?

Yes, but I think I'd prefer to do that in a different PR after this one has been merged.

drmason789 commented:
@cagnulein I've added some unit tests for the lockscreen setup. 

I've added a class of static members "ObjectFactory" which can be used to create objects of currently 2 interfaces, most notably a QZLockscreen object. In IOS this is going to be an instance of the existing lockscreen class, but can be overridden with a delegate, as I have done in the test project, where I instantiate a TestLockscreen object instead, which records the configuration requested (bike, treadmill, rower) and zwiftMode (false when requested by bluetoothdevice, true when requested by virtualdevice).

The LockscreenFunctions object contains the logic for updating the lockscreen for non-Peloton workaround updates, and for deteriming from settings if the Peloton workaround should be activated. This class uses the ObjectFactory to create the QZLockscreen object.

This LockscreenFunctions class I consider to be unfinished. It stores its config more (bike, treadmill, rower) but still has separate methods with the same parameters for the crank/heartrate update. These could become one that uses this config.
There is a separate config of the same data type  in the QZLockscreen interface. This corresponds to the virtual device config that has been set for the Peloton workaround. Currently this would usually be the same, except for the rower, where the lockscreen bike mode is used when not in zwift mode, and the rower mode when used in zwift mode.

I'm having trouble with the ComputrainerBike test crashing the whole test run, maybe because it's calling virtual methods from the constructor (unlike other classes it is doing its virtual device setup in the constructor).

I'm also finding that in some tests, the test lockscreen object is ending up with zwift mode set unexpectedly, which makes me think there's a virtualdevice being created and reconfiguring the lockscreen. This does make me think, in the code prior to and within this PR, with the way the bluetooth class creates bluetoothdevice objects for all types of device it finds, instead of maintaining only 1 instance, and that there is only 1 lockscreen object, does the virtual device or Peloton workaround code for a device not being used by QZ interfere with the lockscreen config for the device that QZ is using?

cagnulein commented:
> I'm having trouble with the ComputrainerBike test crashing the whole test run, maybe because it's calling virtual methods from the constructor (unlike other classes it is doing its virtual device setup in the constructor).

If you want you can modify this, using the same logic did for others bike. I put it in the constructor, because, since it's a serial port device, there is no need to wait a Bluetooth connection before creating a virtual peripheral Bluetooth

drmason789 commented:
I may have interference from dircon. Clearly I have more to do.

![image](https://user-images.githubusercontent.com/29954900/221433727-72b91f85-ffe8-4334-ac02-03cdeb7dec20.png)

drmason789 commented:
I moved the virtual device setup in Computrainerbike, but it's still crashing the tests, and the debugger. If one removes it from BluetoothDeviceTestDataTypes in devices.h, the other tests can be used.

cagnulein commented:
ok leave this out of the game, it's for sure a special case

Il giorno lun 27 feb 2023 alle 04:15 David Mason ***@***.***>
ha scritto:

> I moved the virtual device setup in Computrainerbike, but it's still
> crashing the tests, and the debugger. If one removes it from
> BluetoothDeviceTestDataTypes in devices.h, the other tests can be used.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cagnulein/qdomyos-zwift/pull/1241#issuecomment-1445628737>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALYWBCLKJNIUFKPEGEACTWZQL55ANCNFSM6AAAAAAURPPLIY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
-- 
Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud

drmason789 commented:
@cagnulein Are you able to tell me if this branch builds in IOS?

drmason789 commented:
> @cagnulein Are you able to tell me if this branch builds in IOS?

More specifically, builds _for_ IOS.

cagnulein commented:
sorry @drmason789 i didn't see this message. i will check this asap

cagnulein commented:
@drmason789 for the #1330 i introduced a new function in order to read the AW HR or the Garmin HR. So I will try to change this PR accordingly to the #1330 
Since you already did a lot it's not fair to leave you alone in this PR, do you agree?

drmason789 commented:
> @drmason789 for the #1330 i introduced a new function in order to read the AW HR or the Garmin HR. So I will try to change this PR accordingly to the #1330 Since you already did a lot it's not fair to leave you alone in this PR, do you agree?

Yes.

I've not done much on this lately due to travel and work. 

If I remember correctly, I have 3 tests failing because the lockscreen functions are not being configured as I expect, but how this should be may not be well defined, and 3 tests failing because the devices they cover work differently to the others.

Also, now that you've grouped "lockscreen" with other things in a "external device" concept (something I have in this PR with the IOS lockscreen and the test lockscreen), perhaps it would be appropriate to shift some of the terminology I've used in type names and else where from using LockScreen (e.g. LockscreenFunctions) to something more generic, e.g. ExternalMonitor.

drmason789 commented:
Not ready for merging.

cagnulein commented:
@drmason789 If you update from master you can now check if the iOS builds. I added it to the CI:)

drmason789 commented:
@cagnulein 

I'm wondering if there's any reason you haven't put the "keep awake helper" heart rate code into update_hr_from_external too. 

```
#ifdef Q_OS_ANDROID
    if (settings.value(QZSettings::ant_heart, QZSettings::default_ant_heart).toBool())
        Heart = (uint8_t)KeepAwakeHelper::heart();
    else
#endif
    {
        if (heartRateBeltName.startsWith(QStringLiteral("Disabled"))) {
            update_hr_from_external();
        }
    }
```
update_hr_from_external could do all the checking if ant_heart is enabled, and if the heartRateBeltName starts with "Disabled".

cagnulein commented:
@drmason789 at the beginning I did that, but while I was merging I saw that doing everything inside the `update_hr_from_external` would change the behaviour (because in some class the ant heart thing is linked to the heart sensor of the device)..So in order to do it safely I just leave the things as they are

drmason789 commented:
> @drmason789 at the beginning I did that, but while I was merging I saw that doing everything inside the `update_hr_from_external` would change the behaviour (because in some class the ant heart thing is linked to the heart sensor of the device)..So in order to do it safely I just leave the things as they are

In that case there could be a function "useKeepAwakeHelper()" that returns a value set by each bluetoothdevice subclass constructor, or a virtual function that defaults to the usual value and is overridden by a minority of subclasses, which is then used in update_hr_from_external to determine if the keep awake help code should be used.

cagnulein commented:
do you mean like a 

```
uint8_t useKeepAwakeHelper() {
#ifdef Q_OS_ANDROID
    if (settings.value(QZSettings::ant_heart, QZSettings::default_ant_heart).toBool()) {
        Heart = (uint8_t)KeepAwakeHelper::heart();
        return 1;
    }
    else
        return 0;
#endif
    return 0;
}
```

if so, Yes i guess it's reasonable!

drmason789 commented:
That's not quite what I was thinking. 
I had a look around the code further, and I think the virtual function I mentioned could be avoided.

Currently I would propose changing the definition of bluetoothdevice::update_hr_from_external to take in an optional default value from the machinery, which defaults to 0 (not present). 

The code in update_hr_from_external would be replaced with the whole if structure, then the current code in that function inlined at the new location.

If a default from the machinery is provided, the heart_ignore_builtin check would be done to see if it should be used.

```
void bluetoothdevice::update_hr_from_external(uint8_t defaultFromMachinery=0) {
    QSettings settings;

#ifdef Q_OS_ANDROID
    if (settings.value(QZSettings::ant_heart, QZSettings::default_ant_heart).toBool())
        Heart = (uint8_t)KeepAwakeHelper::heart();
    else
#endif
    {
        QString heartRateBeltName = settings.value(QZSettings::heart_rate_belt_name, QZSettings::default_heart_rate_belt_name).toString();
        if (heartRateBeltName.startsWith(QStringLiteral("Disabled"))) {
            uint8_t heart = 0;

            if(defaultFromMachinery!=0) {
                bool disable_hr_frommachinery = settings.value(QZSettings::heart_ignore_builtin, QZSettings::default_heart_ignore_builtin).toBool();

                if(!disable_hr_frommachinery)
                    heart = defaultFromMachinery;
            }

            if (heart == 0) {
                if(settings.value(QZSettings::garmin_companion, QZSettings::default_garmin_companion).toBool()) {
#ifdef Q_OS_ANDROID
                    Heart = QAndroidJniObject::callStaticMethod<jint>("org/cagnulen/qdomyoszwift/Garmin", "getHR", "()I");
#endif
#ifdef Q_OS_IOS
#ifndef IO_UNDER_QT
                    lockscreen h;
                    Heart = h.getHR();
#endif
#endif
                    qDebug() << "Garmin Companion Heart:" << Heart.value();
                } else {
#ifdef Q_OS_IOS
#ifndef IO_UNDER_QT
                    lockscreen h;
                    long appleWatchHeartRate = h.heartRate();
                    h.setKcal(KCal.value());
                    h.setDistance(Distance.value());
                    Heart = appleWatchHeartRate;
                    qDebug() << "Current Heart from Apple Watch: " << QString::number(appleWatchHeartRate);
#endif
#endif
                }
            } else
                Heart = heart;
        }
    }

}
```

In the simple cases such as chronobike:

```
#ifdef Q_OS_ANDROID
    if (settings.value(QZSettings::ant_heart, QZSettings::default_ant_heart).toBool())
        Heart = (uint8_t)KeepAwakeHelper::heart();
    else
#endif
    {
       QString heartRateBeltName = settings.value(QZSettings::heart_rate_belt_name, QZSettings::default_heart_rate_belt_name).toString();
        if (heartRateBeltName.startsWith(QStringLiteral("Disabled"))) {
            update_hr_from_external();
        }
    }
```

becomes:

`this->update_hr_from_external();`

In the cases with an HR from the machinery such as domyosbike:

```
    QString heartRateBeltName = settings.value(QZSettings::heart_rate_belt_name, QZSettings::default_heart_rate_belt_name).toString();
    bool disable_hr_frommachinery = settings.value(QZSettings::heart_ignore_builtin, QZSettings::default_heart_ignore_builtin).toBool();

#ifdef Q_OS_ANDROID
    if (settings.value(QZSettings::ant_heart, QZSettings::default_ant_heart).toBool())
        Heart = (uint8_t)KeepAwakeHelper::heart();
    else
#endif
    {
        if (heartRateBeltName.startsWith(QStringLiteral("Disabled"))) {
            uint8_t heart = ((uint8_t)value.at(18));
            if (heart == 0 || disable_hr_frommachinery) {
                update_hr_from_external();
            } else
                Heart = heart;
        }
    }

```
becomes

`this->update_hr_from_external((uint8_t)value.at(18));`

drmason789 commented:
@cagnulein I've been looking further through the code and I can see there are a few more cases, e.g. different strategies for the avoidance of the heart rate from the machinery, use of heart rate flags (as a flag and a heart rate), reading the heart rate and not doing anything with it, only doing the heart rate update if there is no virtual device. I'll come back to this later.

drmason789 commented:
@cagnulein I want to do some significant rework on this PR as I think after the Garmin update now there is a clearer separation of external HR device and lockscreen Peloton workaround, and I think the tests should reflect that.

drmason789 commented on tst/Devices/UltrasportBike/ultrasportbiketestdata.h:
```diff
@@ -1,12 +1,12 @@
 
 public:
-    UltrasportBikeTestData() : BluetoothDeviceTestData("Ultrasport Bike") {
+    UltrasportBikeTestData() : BikeTestData("Ultrasport Bike") {
```
@cagnulein In this PR that I have abandoned for later significant rework, I made each test data class inherit from BikeTestData, EllipticalTestData or whichever in order to use inheritance.  With your approval, I would like to make this change as a separate PR.

cagnulein commented:
sure thanks!

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno sab 10 feb 2024 alle 21:37 David Mason ***@***.***>
ha scritto:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In tst/Devices/UltrasportBike/ultrasportbiketestdata.h
> <https://github.com/cagnulein/qdomyos-zwift/pull/1241#discussion_r1485261200>
> :
>
> >
>  public:
> -    UltrasportBikeTestData() : BluetoothDeviceTestData("Ultrasport Bike") {
> +    UltrasportBikeTestData() : BikeTestData("Ultrasport Bike") {
>
> @cagnulein <https://github.com/cagnulein> In this PR that I have
> abandoned for later significant rework, I made each test data class inherit
> from BikeTestData, EllipticalTestData or whichever in order to use
> inheritance. With your approval, I would like to make this change as a
> separate PR.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cagnulein/qdomyos-zwift/pull/1241#pullrequestreview-1874205354>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALYWBBXUPVISWBQ3QGXCTYS7LBJAVCNFSM6AAAAAAURPPLI2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQNZUGIYDKMZVGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
